### PR TITLE
Starting Typescript Migration

### PR DESCRIPTION
**Pull request by herbertli**
- Added typescript packages to package.json
- Migrated styles.js (changed ext.)
- Migrated firebase.js (changed ext.)
- Migrated constants.js (changed ext.)
- Migrated GridListCustom.js
- Migrated SpecTile.js
- Migrated SpecList.js
- Migrated Piece.js
- Migrated PieceList.js
- Migrated ItemTypes.js
- Migrated ElementTile.js
- Migrated ElementList.js
- Migrated BoardTile.js
- Migrated BoardList.js
- Migrated Logout.test.js
- Migrated Logout.js
- Migrated NavigationMaster.js
- Migrated Sidebar.js
- Migrated Board.js
- Migrated CanvasImage.js
- Migrated Element.js
- Migrated SpecInfo.js
- Migrated SpecInfo.test.js

SpecViewer still needs migrating, but react-contenteditable has no typescript support

@yoav-zibin @apsdehal Sorry, I couldn't figure out how to add a reviewer.
Please let me know if I need to change anything!

### CODE REVIEW

File: src/components/GridListCustom.tsx [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 
 import { GridList } from 'material-ui/GridList';
 import Subheader from 'material-ui/Subheader';
 import styles from '../styles';
+import { CSSProperties } from 'react';
 
-function GridListCustom(Component) {
-  return props => {
+function GridListCustom(Component: React.StatelessComponent | React.ComponentClass) {
+  return (props: any) => {
```
Comment (apsdehal) [inline src/components/GridListCustom.tsx:10]: Maybe you can move away from any here also.

File: src/components/gamespec/Board.tsx [hunk 2]
```diff
@@ -274,7 +306,7 @@ class Board extends React.Component {
                   onClick={() => {
                     this.handleClickOn(index);
                   }}
-                  item={item}
+                  // item={item}
```
Comment (apsdehal) [inline src/components/gamespec/Board.tsx]: This should not be commented I think?

File: src/components/gamespec/BoardTile.tsx [hunk 3]
```diff
@@ -1,10 +1,17 @@
...
 
 import styles from '../../styles';
 
-const BoardTile = props => {
+interface BoardTileProps {
+  keyProp: any;
+  selectedKey: any;
+  handleGridTileClick: (keyProp: any, image: any) => void;
+  image: any;
+}
```
Comment (apsdehal) [inline src/components/gamespec/BoardTile.tsx:12]: Using any doesn't provide any benefits of TypeScript. If possible, can you move away from any?

**Review by apsdehal — CHANGES_REQUESTED** [review]
Overall, looks good. With some `any` types which can be possibly removed.

Can you do one more thing, if possible can you rebase ts-migrate branch with current master as there are some changes I made there.

**Comment by apsdehal** [comment]
Also, try not to push big PRs. It is hard to review and harder to rebase in case some change happens in master. Use small commits, so that they are easy to rebase with new upstream changes. Now, it will be a lot of pain and effort to rebase with master.

File: src/components/gamespec/Board.tsx (hunk 2 above)
Comment (herbertli) [reply src/components/gamespec/Board.tsx]: Just wanted to clarify... is item={item} passing down all other props down to the child component, or is there an actual prop called "item"? Thanks!

Comment (apsdehal) [reply src/components/gamespec/Board.tsx]: It is getting passed down, it is used somewhere, but I don't remember where.

Comment (apsdehal) [reply src/components/gamespec/Board.tsx]: Got it, it is used by react-dnd when we drag and drop.

**Comment by herbertli** [comment]
@apsdehal I think I've managed to get my branch in sync with master, is there a way to test the site locally?

**Comment by apsdehal** [comment]
Can you follow the instructions in Readme for setup? Once you have everything I can provide you credentials for Firebase.

**Comment by apsdehal** [comment]
@herbertli Were you able to run it?

**Comment by herbertli** [comment]
@apsdehal it compiles correctly, just need authentication credentials!

**Comment by apsdehal** [comment]
Check piazza

**Comment by apsdehal** [comment]
@herbertli Consider this complete (make sure there aren't any files left if you are continuing with this homework). I will need some time to pull down rebase and then push stuff.

Thanks.

**Comment by apsdehal** [comment]
Ok, I have merged this into master. You can delete your ts-migrate branch and create a new one from master.

**Pull request closed without merging**